Move store credits in core

IrvinFan opened the pull request:
This is the starting point of merging store credits feature (but leave gift cards feature) from extension https://github.com/paulo-surfdome/spree_store_credit_payment_method/tree/move_credit_out into Spree.

Here is the approach we took: go through codes in the directories in https://github.com/paulo-surfdome/spree_store_credit_payment_method/tree/move_credit_out and copy (new file) or merge (decorator) store credits related codes into Spree. And at the same time remove those codes from the extension (Related PR: https://github.com/spree-contrib/spree_store_credit_payment_method/pull/24).

Right now the key components are done. Anyone can take from here.
- [x] merge app directory
- [x] merge bin directory
- [x] merge db directory
- [x] merge config directory
- [x] merge lib directory
- [x] merge spec directory
- [x] implement store frontend UI

ifan-godaddy committed:
Move store credits in core

JDutil commented:
We'll need to get this finished up soon to be included in Spree 3.  We're getting near a feature freeze for RC.

IrvinFan commented:
I'll clean this up tonight. But Joe was volunteering to help build the store front (last checkbox item in this PR). Now sure how that goes.

JDutil commented:
Ping @simmsy any updates on views for this?

JDutil commented:
Looks like some views are submitted in a PR here https://github.com/spree-contrib/spree_store_credit_payment_method/pull/10 that can be used.

paultyng commented:
If you intend to pull that in, it would be nice to fix the issue of it asking you to supply payment information even if you have enough store credit to cover the order.  I've fixed it on our site and its live, but it was a little hacky as it requires some core changes, but if this is merging in, that's probably doable.

gmacdougall commented:
@paultyng I have also fixed the issue where it requires payment information, but it is also a terrible hack.

Something along the lines of:

```Ruby
    def add_store_credit_payments
      existing_credit_card_payment.destroy if covered_by_store_credit?
      super
    end
```

paultyng commented:
I had to make a number of modifications due to other issues as well (I do not have a confirm screen as a limitation of my payment method, paypal):

* Add store credit payments before transition to `:payment` state (and change `total_applicable_store_credit` to use payments sum in the `:payment` state as well, but this did have the benefit of letting me remove the reconcile with credit card code)
* Change `update_params_payment_source` to use store credit information in its calculation of amount
* Update the paypal controller to deal with store credit in that extension
* Change the **spree/checkout/_payment.html.erb** view to have a dummy hidden field so that the required params were satisfied

But it wasn't that bad.  The main difference between my code and the stuff in the existing extension was the move to the before payment transition due to my missing confirm state.

peterberkenbosch commented:
Working on this today and the next if needed to get this in. Focus on the logic first, views later.

ifan-godaddy committed:
Copy view codes from https://github.com/spree-contrib/spree_store_credit_payment_method/pull/10

ifan-godaddy committed:
Add the hack to fix the issue of it asking you to supply payment information even if you have enough store credit to cover the order

IrvinFan commented:
Sorry for taking so long to work on this. I finally find some time today and I have done the following:

1. Moved store credit codes from https://github.com/spree-contrib/spree_store_credit_payment_method to Spree.
2. Include latest sotre credit codes in this PR: https://github.com/spree-contrib/spree_store_credit_payment_method/pull/23
3. Copied view codes from this PR: https://github.com/spree-contrib/spree_store_credit_payment_method/pull/10
4. Added @gmacdougall [hack](https://github.com/spree/spree/pull/5625?w=1#issuecomment-69360178)

I didn't look into @paultyng [solution for the same issue](https://github.com/spree/spree/pull/5625?w=1#issuecomment-69361879) Please feel free to submit a PR to this branch.

Keep in mind that I barely moved/copied the codes. Haven't got a chance to actually try the codes. I won't be surprised if rspec and the funstionality are broken. I'll keep working on fixing them. Just think I would like to give you guys a head up that this task is now in "bug fixing" mode. Anybody would like to help test/fix are welcome. Please let me know if you need direct write permision to this branch.

Thanks

gmacdougall commented:
Could you please pull the mention of me out of the commit message? If it's merged like that, GitHub will e-mail anytime that commit gets cherry-picked or merged in to another branch.

IrvinFan commented:
@gmacdougall ammended the commit message.

athal7 commented on api/app/controllers/spree/api/store_credit_events_controller.rb:
```diff
@@ -0,0 +1,9 @@
+class Spree::Api::StoreCreditEventsController < Spree::Api::BaseController
+  def mine
+    if current_api_user.persisted?
```
this seems like an odd check.

athal7 commented on backend/app/views/spree/admin/shared/_menu.html.erb:
```diff
@@ -7,3 +7,12 @@
   </div>
 </nav>
+
+<% if request.fullpath.match(/^\/admin\/users/) %>
```
hm this seems different from how submenus are usually done.

athal7 commented on core/app/models/spree/order.rb:
```diff
@@ -692,6 +693,18 @@ def after_cancel
       self.update!
+
+      # Free up authorized store credits
+      payments.store_credits.pending.each { |payment| payment.void! }
```
why would we be in a state with pending store credits where we had never previously been concerned about pending payments of another kind?

athal7 commented on core/app/models/spree/order.rb:
```diff
@@ -692,6 +693,18 @@ def after_cancel
+      # Free up authorized store credits
+      payments.store_credits.pending.each { |payment| payment.void! }
+
+      # payment_state has to be updated because after_cancel on
```
this seems odd to introduce in the context of store credits. they ought to be behaving the same as credit card payments, and anything to the contrary should be handled in the payment method, not here imho

athal7 replied:
plus this comment references 'super' which is no longer relevant

athal7 commented on core/app/models/spree/order/checkout.rb:
```diff
@@ -102,6 +102,8 @@ def self.define_state_machine!
               before_transition to: :resumed, do: :ensure_line_item_variants_are_not_deleted
               before_transition to: :resumed, do: :ensure_line_items_are_in_stock
 
+              before_transition to: :confirm, do: :add_store_credit_payments
```
we probably have to do this somewhere unrelated to confirm since it is an optional state

JDutil commented:
@athal7 theres a newer branch here if you could review there sorry https://github.com/spree/spree/pull/5917
